### PR TITLE
[RLlib] Examples folder cleanup vol. 45: Enhance/redo autoregressive action distribution example.

### PR DESCRIPTION
**Pull request by sven1977**
Examples folder cleanup vol. 45: Enhance/redo autoregressive action distribution example.
* Slightly changed the correlated-actions-env to make sure a1 is dependent on obs and a2 is dependent on both obs AND a1.
* Simplified example RLModule by removing all catalog dependencies and the encoder.
* Fixed bug in example RLModule: discrete a1 was not one-hot'd, which was probably the reason why learning was not working or very slow.
* This PR is also in preparation of another upcoming doc PR that explains the various action distribution types available to users.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/autoregressive_action_dist.py [hunk 1]
```diff
@@ -1,223 +0,0 @@
-# @OldAPIStack
```
Comment (sven1977) [inline rllib/examples/autoregressive_action_dist.py:1]: old API stack -> remove

File: rllib/examples/envs/classes/correlated_actions_env.py [hunk 2]
```diff
@@ -1,74 +1,79 @@
...
-    other cases the negative deviance between the desired action a2 and its
-    actual counterpart serves as reward. The reward is constructed in such a
-    way that actions need to be correlated to succeed. It is not possible
-    for the network to learn each action head separately.
+    There are two reward components:
+    The first is the negative absolute value of the delta between 1.0 and the sum of
+    obs + a1. For example, if obs is -0.3 and a1 was sampled to be 1, then the value of
+    the first reward component is:
+    r1 = -abs(1.0 - [obs+a1]) = -abs(1.0 - (-0.3 + 1)) = -abs(0.3) = -0.3
+    The second reward component is computed as the negative absolute value
```
Comment (simonsays1980) [inline rllib/examples/envs/classes/correlated_actions_env.py:21]: Maybe  better to keep the same values as used for the first reward component.

File: rllib/examples/envs/classes/correlated_actions_env.py [hunk 3]
```diff
@@ -1,74 +1,79 @@
...
-        # Extract actions
+        # Extract individual action components, a1 and a2.
         a1, a2 = action
+        a2 = a2[0]  # dissolve shape=(1,)
 
-        # The state determines the desired relationship between a1 and a2
-        desired_a2 = (
-            self.state[0] * a1
-        )  # Autoregressive relationship dependent on state
+        # r1 depends on how well q1 is aligned with obs:
```
Comment (simonsays1980) [inline rllib/examples/envs/classes/correlated_actions_env.py:68]: q1 -> a1 ;)

File: rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py [hunk 4]
```diff
@@ -1,203 +1,135 @@
...
+            Columns.ACTION_DIST_INPUTS: torch.cat([prior_out, posterior_out], dim=-1),
+            # Concatenate the prior and posterior actions and log probabilities.
+            Columns.ACTIONS: actions,
+        }
+
+        return outputs
+
+    @override(TorchRLModule)
+    def get_inference_action_dist_cls(self):
+        return TorchMultiDistribution.get_partial_dist_cls(
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py:132]: This is a nice example of how to use our `MultiDistribution` classes. I wonder, however, why we do not use it above where the distributions are used for sampling? In training I guess exactly this distribution is used for the action logps, isn't it?

File: rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py [hunk 5]
```diff
@@ -1,203 +1,135 @@
...
-            hidden_layer_activation=self.model_config["head_fcnet_activation"],
-            output_layer_dim=self.required_output_dims[0],
-            output_layer_activation="linear",
+        # Assert the action space is correct.
+        assert isinstance(self.action_space, gym.spaces.Tuple)
+        assert isinstance(self.action_space[0], gym.spaces.Discrete)
+        assert self.action_space[0].n == 3
+        assert isinstance(self.action_space[1], gym.spaces.Box)
+
+        self._prior_net = nn.Sequential(
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py:44]: This new definition of the autoregressive module simplifies the setup and is better readable, I admit. However, it completely blanks out how users could/should use the `RLModule` attributes as well as the `Catalog` that makes setting up networks simply by pre-configured hyparparameters. Given that most current examples define the RLModule directly without leveraging its specific attributes, methods, or the `Catalog` - and recognizing that this often presents a challenge - we should evaluate our approach. Specifically, we should consider whether to (a) provide these same examples also in a form that explicitly demonstrate the use of these class attributes, methods, and the catalogue, or (b) streamline the `RLModule` by simplifying it further and potentially removing the `Catalog` entirely.

File: rllib/examples/envs/classes/correlated_actions_env.py [hunk 6]
```diff
@@ -1,74 +1,79 @@
...
+    The first is the negative absolute value of the delta between 1.0 and the sum of
+    obs + a1. For example, if obs is -0.3 and a1 was sampled to be 1, then the value of
+    the first reward component is:
+    r1 = -abs(1.0 - [obs+a1]) = -abs(1.0 - (-0.3 + 1)) = -abs(0.3) = -0.3
+    The second reward component is computed as the negative absolute value
+    of `obs + a1 + a2`. For example, if obs is 0.5, a1 was sampled to be 0,
+    and a2 was sampled to be -0.7, then the value of the second reward component is:
+    r2 = -abs(obs + a1 + a2) = -abs(0.5 + 0 - 0.7)) = -abs(-0.2) = -0.2
+
+    Because of this specific reward function, the agent must learn to optimally sample
```
Comment (simonsays1980) [inline rllib/examples/envs/classes/correlated_actions_env.py:26]: Maybe describe in a sentence why this specific reward design supports the agent in learning correlated actions?

File: rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py [hunk 7]
```diff
@@ -1,203 +1,135 @@
...
+
+        # If in inference mode, we need to set the distribution to be deterministic.
+        if inference:
+            dist_a1 = dist_a1.to_deterministic()
+        # Sample a1.
+        a1 = dist_a1.sample()
+
+        # Posterior forward pass.
+        posterior_batch = torch.cat(
+            [obs, one_hot(a1, self.action_space[0])],
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py:106]: Interesting. I tried in multiple examples both, `one-hot` and "raw" actions and could not see a difference. HOw many iterations does it need now to converge?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Some general questions in regard to the simplification of the `RLModule` examples. Otherwise small nits.

File: rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py (hunk 4 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py:132]: It's still a very messy API:
* `from_logits` requires a single tensor, which then gets split into n sub-tensors through the help of another method 🤔 
* `ctor` however utilizes a nested structure of subcomponents, which should probably be the norm. I'm wondering, whether we can simplify the entire `Distributions` API further by getting rid of `from_logits` altogether.
* Another, even more radical thought would be to even get rid of `Distribution/TorchDistribution` and use `torch.Distribution` instead 😉.

But we should clean this up separately.

File: rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py (hunk 5 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py:44]: By adding too many APIs to this example (Catalog) that have nothing to do with the actual problem to be solved here (autoregressive sampling of the 2 action components), we are cluttering this example.

I agree, we need to think about the usefulnes of `Catalog` in general. My take has been for a while now to get rid of it and simply replace it with utility functions that the user may use, but doesn't have to. Most users don't have the problem of 1 Model -> 100 spaces (like RLlib does!), but have a relatively fixed space and thus also a relatively stable model encoder architecture that doesn't require a complex "catalog decision tree".

File: rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py (hunk 7 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py:106]: Yeah, I think the existing example was not learning or just super slowly. The new example definitely learns both action components' dependency on 1) obs and 2) obs+[first action].
